UIPS/UIP-0014.md

```diff
@@ -0,0 +1,32 @@
+---
+uip: 0114
+title: ?_ rune, noun equality restricts type to left parameter
+description: for use with dependently typed maps and other related value-based type inference
+author: ~sarpen-laplux (@xiphiness)
+status: Draft
+type: Standards Track
+category: Kernel
+created: ~2023.8.9
+---
+
+## Abstract
+
+This PR introduces a new rune ?_(a=hoon b=wing), subject to renaming, whose compiled nock is equivalent to .=, but refines the type of b to that of a on noun equality.
+
+## Specification
+
+See https://github.com/urbit/urbit/pull/6656
+
+## Rationale
+
+This procedure can be generalized to work with arbitrary ships, but it's not necessary that it be used for anyone other than our galaxy until we either start forwarding through stars, or need some keep-alive mechanism for "sticky" scry requests. Both of those are planned for the future. The current design keeps formal pings from /app/ping to the rest of the (non-galaxy) sponsorship hierarchy. These could be removed, since galaxies can statelessly forward packets from the sponsor star to the sponsee.
+
+Initially only galaxies will need to act as STUN servers, but any ship could support STUN requests. These could be used to check connectivity with other ships outside of the event loop, and, if additionally listening on a standard port, could provide useful service to applications running outside of urbit (such as webrtc clients).
+
```

Comment:
STUN with `?_`!

Reply:
it's a very cool, entirely unexpected feature!